### ISSUE TITLE
Documentation: what is the semantics of getNBestList

### ISSUE THREAD

**Issue by Avmb** [marian-nmt/marian-dev#334]
As used here: 
https://github.com/marian-nmt/marian-dev/blob/86103bea6a5a5a28b413d48862a6b947824401e6/src/translator/beam_search.h#L274

I'm trying to use getNBestList to extract the top-k word scores (not path scores) for each state in the beam and batch, but I can't quite understand what all the parameters mean and what this function actually does. Specifically, what is beamSizes and what is the format of the output vectors (the pathScores tensor is 4-d, while the output vectors are flat).

I want to give it a tensor and get the top-k indices and scores over the vocabulary dimension (the fourth of the pathScores tensor, if I understand correctly).

**Comment by frankseide** [comment]
Unfortunately I don't have time for a detailed response, but please have a look at `src/microsoft/quicksand.cpp`, where `decode()` interprets the returned N-best list. You may be able to deduce from this code what you need. (This is the code that interfaces the Marian decoder with our production system, and is not compiled as part of regular Marian.)

**Comment by frankseide** [comment]
Re-reading your question, you actually don't need the outer API, but the inner function. From reading code, it seems that `beamSizes` is an unnecessarily complex way of passing the scalar `localBeamSize` to `getNBestList()`. `beamSizes` is an array indexed by the batch index, created once filled with `localBeamSize` in all of its elements, and then passed as a `const` reference and not ever again changed. So one could just pass `localBeamSize` as a scalar.

If I understand it right, `localBeamSize` starts out as the beam-size parameter, but as decoding progresses and more and more beam entries across the batch reach the end of sentence, the beam size reduces over time.

The sizes of the two returned vectors is equal to #batch entries * beam size. `outPathScores` contains accumulated partial path scores up to the current decoding step. `outKeys` contains indices. From reading the inner `getNBestList()` function, both seem to be indexed as this: `[batch index * localBeamSize + i]` where `i=0..localBeamSize -1` indexes one of the top `localBeamSize ` hypotheses sorted by partial path score (highest first). These are not word scores (softmax values).

The word scores are in `scores = pathScores->val()`, where `scores->data()` gives you a flattened version of the tensor. I don't have the tensor shape ready, but if you know it, the tensors are in fastest-changing-dimension-last layout.

If you still need to know what the `outKeys` are, let me know, need to dig.

And please do let me know if you find any of the above to be incorrect; I just derived that from looking at the code, it may be wrong.

**Comment by emjotde** [comment]
Just injecting myself here quickly. That has nothing (or only tangentially) to do with the actual n-best list, Frank is talking about a different piece of code. That's the top-k search in the beam, it's just horribly named for no good reason and that's where the misunderstanding here comes from. 

I will comment some more later today.

**Comment by emjotde** [comment]
In case you saw my previous comment in your e-mail inbox. I deleted it here as Frank's explanation was correct.

**Comment by Avmb** [comment]
Hi Frank and Marcin, thanks for the answers.

I'm still trying to figure out how to use getNBestList() to essentially replicate the functionality of Tensorflow's tf.nn.top_k() (though I don't care about differentiability, since I'm still using it only during decoding).

If I understand correctly, getNBestList() takes a 4-d tensor with dimensions [dimBatch, 1, beam_size, vocabSize] and computes the top-k over both the third and forth dimension (and possibly the second, but I don't know since it's always equal to 1). I'm trying to compute the top-k over only 4th dimension (the vocabulary), so I reshape the tensor to make it [dimBatch* beam_size, 1, 1, vocabSize] and I pass it together with a beamSizes initialized with the (dimBatch*beam_size, k) constructor, where k = localBeamSize (I would also be able to use other values for k, but it's not crucial). 

But it works only for the first word, where the effective beam size is equal to 1 (though, confusingly, localBeamSize is still equal to 12), at the second word I get:

Error: invalid argument - /fs/meili0/amiceli/marian-word-probs/src/translator/nth_element.cu:332

The shapes of the things involved are: 
scores->shape(): shape=24x1x1x50000 size=1200000, beamSizes.size(): 24, beamSizes[0]: 12, dimBatch: 2, localBeamSize: 12

Is it possible to use getNBestList() as a generic top-k function? What is the meaning of the parameter "first"? What is the format of the outKeys? (I suppose outKey[i] = beam_size * beam_index + vocab_index)

**Comment by frankseide** [comment]
I would not try to shoe-horn this into `getNBestList()`, but rather code it yourself.

>The word scores are in scores = pathScores->val(), where scores->data() gives you a flattened CPU-side version of the tensor.
> 4-d tensor with dimensions [dimBatch, 1, beam_size, vocabSize]

I would just work from this, and possibly steal code from `getNBestList()` where useful (MIT license has to be good for something!).

To implement the top-k over the last axis as I understand from your description, I think you would need to sort the elements separately for each batch and beam entry. That should be not too hard since the elements of the last axis are already consecutive in memory. So you just need to compute the slice range and pass it as begin and end iterators to std::sort.

**Comment by Avmb** [comment]
I wanted to use getNBestList() because it can run on the gpu, I would like to avoid writing my own CUDA kernel if I can. Do you know if there is a big performance loss from running the top-k operation on the cpu?

**Comment by frankseide** [comment]
I can't see how the current version runs on the GPU. Are you sure that it does?

**Comment by Avmb** [comment]
I think so: https://github.com/marian-nmt/marian-dev/blob/86103bea6a5a5a28b413d48862a6b947824401e6/src/translator/beam_search.h#L174

**Comment by frankseide** [comment]
Ah... I was looking at the CPU version. I think what I would do is run your code on the CPU and single-step through the code to debug it. Once it works, then switch to GPU.

**Comment by snukky** [comment]
The beam search implementation has changed a lot since the issue was open. Closing.
